File: frontend/src/api/courseBulletin.ts
import axios from "axios";
import {
    CourseBulletin
} from "../schemas/courseBulletin";

export async function getCourseBulletinList(course_id:number) :Promise<Array<CourseBulletin>>{
    let url = "/course/bulletin/particular_course/"+course_id;
    try {
        const response = await axios.get(url,{
          });

        const sortByReleaseTime = (a: CourseBulletin, b: CourseBulletin) => {
            if (a.release_time && b.release_time){
                return new Date(b.release_time).getTime() - new Date(a.release_time).getTime();
            }
        };
        const result = response.data;

        const pinedResult = result
            .filter((data: CourseBulletin) => data.pin_to_top)
            .sort(sortByReleaseTime);
        const notPinedResult = result
            .filter((data: CourseBulletin) => !data.pin_to_top)
            .sort(sortByReleaseTime);
        
        return pinedResult.concat(notPinedResult);
    }
    catch(error){
        return[];
    }
}

export async function postCourseBulletin(courseBulletin:CourseBulletin) :Promise<CourseBulletin>{
    let url = "/course/bulletin?uid="+courseBulletin.uid+"&course_id="+courseBulletin.course_id;
    try {
        await axios.post(url,courseBulletin);
    }
    catch(error) {  
        
    }
    return courseBulletin;
    

}

export async function deleteCourseBulletin(id:number){
    let url = "/course/bulletin/"+id;
    try {
        await axios.delete(url,{});
    }
    catch(error) {  
        
    }
}

export async function updateCourseBulletin(courseBulletin:CourseBulletin){
    let url = "/course/bulletin/"+courseBulletin.id;
    try {
        await axios.put(url,courseBulletin);
    }
    catch(error) {  
        
    }
}

File: backend/api/routers/course_bulletin.py
from fastapi import APIRouter, Depends, HTTPException, status

from .depends import check_course_id, check_course_bulletin_id, check_user_id
from crud.course_bulletin import CourseBulletinCrudManager
from crud.notification import NotificationCrudManager
from crud.selected_course import SelectedCourseCrudManager
from schemas import bulletin as BulletinSchema

not_found = HTTPException(
    status_code=status.HTTP_404_NOT_FOUND, 
    detail="Course bulletin does not exist"
)

already_exists = HTTPException(
    status_code=status.HTTP_409_CONFLICT, 
    detail="Course bulletin already exists"
)

CourseBulletinCrud = CourseBulletinCrudManager()
router = APIRouter(
    prefix="/api/course",
    tags=["Course Bulletin"]
)

NotificationCrud = NotificationCrudManager()
SelectedCourseCrud = SelectedCourseCrudManager()

@router.post(
    "/bulletin", 
    status_code=status.HTTP_204_NO_CONTENT
)
async def create_course_bulletin(
    newBulletin: BulletinSchema.BulletinCreate,
    uid: str = Depends(check_user_id),
    course_id: str = Depends(check_course_id)
):
    """
    Create a bulletin with the following information:
    - **release_time**
    - **title**
    - **content**
    - **pin_to_top**
    """
    bulletin = await CourseBulletinCrud.create(uid, course_id, newBulletin)
    users = await SelectedCourseCrud.get_by_course_id(course_id)
    for user in users:
        await NotificationCrud.create(user["uid"], bulletin.id, "course_bulletin")
        
    return bulletin


@router.get(
    "/bulletin/{cb_id}", 
    response_model=BulletinSchema.BulletinReadByID,
    status_code=status.HTTP_200_OK
)
async def get_course_bulletin(cb_id: int):
    """
    Get the course bulletin.
    """
    bulletin = await CourseBulletinCrud.get(cb_id)
    if bulletin:
        return bulletin

    raise not_found


@router.get(
    "/bulletin/particular_course/{course_id}", 
    response_model=list[BulletinSchema.CourseBulletinListRead],
    status_code=status.HTTP_200_OK
)
async def get_all_course_bulletins_in_particular_course(
    uid: str = Depends(check_user_id), 
    course_id: int = Depends(check_course_id)
):
    """
    Get all course bulletins in particular course and the subscription status of user.
    """
    bulletins = await CourseBulletinCrud.get_by_course_id(uid, course_id)
    if bulletins:
        return bulletins
    
    raise not_found
   

@router.put(
    "/bulletin/{cb_id}",
    status_code=status.HTTP_204_NO_CONTENT
)
async def update_course_bulletin(
    updateBulletin: BulletinSchema.BulletinUpdate, 
    cb_id: int = Depends(check_course_bulletin_id)
):
    """
    Update the particular course bulletin with at least one of the following information:
    - **title**
    - **content**
    - **pin_to_top**
    """
    await CourseBulletinCrud.update(cb_id, updateBulletin)
    bulletin = await CourseBulletinCrud.get(cb_id)
    
    users = await SelectedCourseCrud.get_by_course_id(bulletin["course_id"])
    for user in users:
        await NotificationCrud.create(user["uid"], cb_id, "course_bulletin")
    
    return 


@router.delete(
    "/bulletin/{cb_id}",
    status_code=status.HTTP_204_NO_CONTENT
)
async def delete_course_bulletin(cb_id: int = Depends(check_course_bulletin_id)):
    """
    Delete the course bulletin.
    """
    await CourseBulletinCrud.delete(cb_id)
    return 
